### ISSUE TITLE
[data grid] Multi filter with custom data object that are not just array of strings

### ISSUE THREAD

**Issue by kasperfilstrup** [mui/mui-x#4307]
### Order ID 💳

#34405

### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### The problem in depth 🔍

I'm trying to use a filter in DataGridPro on a column where the dataset consists of the following data structure:

```js
options: [{
  id: "107e9741-9260-e611-a33e-000c29311e8e",
  name: "Option name 1",
  description: null,
  endDate: "0001-01-01T00:00:00.000",
  startDate: "0001-01-01T00:00:00.000",
  },
  {
  id: "507e9741-9260-e611-a33e-000c29311e8s",
  name: "Option name 2",
  description: null,
  endDate: "0001-01-01T00:00:00.000",
  startDate: "0001-01-01T00:00:00.000",
}]
```

So my current implementation looks like this:
```tsx
filterOperators: [
  {
    label: 'Indeholder',
    value: 'is',
    InputComponent: NameTypeInput,
  },
],
...

function NameTypeInput(params: GridFilterInputValueProps) {
  return <GridFilterInputMultipleSingleSelect {...params} getOptionLabel={offerTypeName} />;
}
```

however, it renders like this when I select any option:
![image](https://user-images.githubusercontent.com/632686/160604690-c7415bb3-df48-48a4-88b9-80e34719e534.png)

Basically I would like a filter identical to the "country" filter of the demos of DGPX, however, it doesn't play well with my exact data structure and I can't seem to find a useful example of how to go about this. Maybe I'm looking in all the wrong places, but I've tried everything from creating a custom filter input with `Autocomplete` and also just trying to use the exposed `GridFilterInputMultipleSingleSelect` component, but with little to no luck.

Any help / pointers would be very helpful

### Your environment 🌎

`npx @mui/envinfo` <- doesn't work at all in my setup (Typescript, React 17, MUI 5+ on Windows 10)

**Comment by DanailH** [comment]
HEy, @kasperfilstrup thanks for raising this!
! Please provide a minimal reproduction test case with the latest version. This would help a lot 👷 .
A live example would be perfect. [This codesandbox.io template](https://codesandbox.io/s/datagrid-v5-quick-start-54iz1) _may_ be a good starting point. 

That being said the `country` filters use a `singleSelect` column type. Maybe you can convert your column to be of that type? Also if you need a different format of the data you can use `valueFormatter` in that column to modify the data.

I think it will be easier if we can see an example of the problem in order to push the discussion further.

**Comment by kasperfilstrup** [comment]
Hey @DanailH - thanks for getting back to me. I'm very sorry I didn't provide you with a not working sandbox example. 🙏

I've tried to create my scenario by forking the sandbox you sent. What I'm trying to do it to have the filter be a multiple input, but seeing that my solution uses server side data AND sorting AND filtering, I need to map the selected filter values back to the corresponding ID's before calling my server.

I've managed to create a multiple input in the sandbox, however - and as you can see - I'm struggling with getting the `valueOptions` rendered correctly and I'm afraid I've come to a full stop now, code wise. 😫

So any help would be appreciated - also a humble wish for the future of the MUIX docs, would be a more comprehensive guide to server side sorting, filtering, etc. 🙏

https://codesandbox.io/s/datagrid-v5-quick-start-forked-bxx48t?file=/src/App.tsx

**Comment by alexfauquette** [comment]
The problem is You tried to modify you inputs to match our component requirement
```js
.map((x) => ({ value: x.id, label: x.name }));
```

And at the same time, you modify the option rendering of this same component to match your inputs
```js
renderOption={(option: any) => <li id={option.id}>{option.name}</li>} // option.name should be otpin.label
```

**Comment by kasperfilstrup** [comment]
@alexfauquette Thank you very much - it works - also on my machine in my setup 😍
